### PR TITLE
fix(utils): a refusal that cannot quote what it refuses degrades, not raises (closes #1903)

### LINKED ISSUE #1903: name_list_error's mapping refusal quotes the mapping's own keys, and that read can raise - how should a refusal degrade when its remedy is unmeasurable?
Follow-up from #1897, measured while fixing it and deliberately left out of that change - the last member of this family still open.

`utils.name_list_error` refuses a `Mapping` and offers its keys as the remedy:

```python
if isinstance(value, Mapping):
    return (
        f"{context}: {param} must be a list of names, not a mapping, "
        f"got {_refusal_container_repr(value)}. "
        f"A mapping is iterable over its keys, so its values would be discarded - "
        f"pass the names as a list: {_refusal_container_repr(list(value))}."
    )
```

That `list(value)` is a read of the caller's value, evaluated **before** `_refusal_container_repr` is entered, so the renderer's never-raise guarantee does not cover it. A `Mapping` whose key iteration raises escapes:

```
name_list_error(HostileKeyMapping(), "cameras", "render_all")  ->  RuntimeError: no keys for you
```

where

```python
class HostileKeyMapping(Mapping[str, str]):
    def __len__(self): return 1
    def __getitem__(self, key): return "camera"
    def __iter__(self): raise RuntimeError("no keys for you")
```

Measured on `61946f6`. Pinned as a boundary by `TestTheNameListIsReadOnceAndAnsweredNotEscaped::test_a_mapping_whose_keys_cannot_be_read_is_a_stated_boundary`, so closing this has to **replace** that test rather than pass it silently.

## Why it was not folded into #1897

#1897 was about a **read count**: the guard read the value on three unguarded branches, so it now reads it once and computes every verdict from that read. This escape survives that change because it is on a branch reached *before* the `Sequence` check, and it is not the same question.

For every other route in the family, the escape is that the verdict itself could not be reached. Here the verdict is already known and is not in doubt - **a mapping is refused whatever its keys say** - and only the *remedy* is unmeasurable. So closing it is not a matter of guarding a read; it is a decision about what a refusal says when it cannot compute the advice it wants to give, and there is no precedent in `utils.py` to follow.

## The decision this needs

Three candidates, none obviously right:

1. **Answer with the read failure**, dropping the mapping verdict. Honest and adds no vocabulary - it is the text #1897 already introduced - but it *understates*: it stops telling the caller they passed a mapping, which is the mistake actually being made. Measured behaviour if the branch simply routes through `_read_name_list`.
2. **Keep the mapping verdict and degrade the remedy**, e.g. "pass the names as a list; its own keys could not be read to quote them here (`RuntimeError: no keys for you`)". Strictly the most informative, and the only option that states both what was measured and what was not - at the cost of a new message form.
3. **Keep the verdict and drop the remedy clause** when the keys cannot be read. Smallest text, but silently omits advice on one path and says nothing about why.

Whichever is chosen, note that `_refusal_container_repr(value)` earlier in the same message already renders the mapping safely (it falls back per component and its fallback is itself guarded), so the mapping verdict is available on this path - the escape is only in the remedy.

## Scope

- One branch of one guard, `utils.name_list_error`.
- `tests/test_container_refusals_render_elementwise.py` owns this family's assertions; the boundary pin named above is the test to replace.
- Worth checking in the same pass whether any other refusal in `utils.py` reads a caller value to build its *remedy* rather than its verdict. `TestNoGuardRendersACallerValueDirectly` scans for unrendered values, not for reads, so an unguarded `list(...)` feeding a renderer is exactly what that scan is blind to - which is how this one survived #1873 and #1875.

## Family

Rendering (#1873), scalar conversion (#1874), container conversion (#1875), `__iter__` (#1878), the shared length probe (#1888), element production (#1889), the name-list read count (#1897). Each is the same defect - a guard whose entire purpose is to answer an unusable input with a structured refusal, raising instead - on a different path.

### PR DESCRIPTION
**Pull request by cagataycali**
Closes #1903.

`utils.name_list_error` built its refusal text by reading the caller's value, and those reads were unguarded. The mapping branch offered the mapping's own keys as its remedy:

```python
f"pass the names as a list: {_refusal_container_repr(list(value))}."
```

The *render* is safe; building its argument runs the caller's `__iter__`. So the guard whose entire purpose is to answer an unusable input with a structured message raised instead:

```
name_list_error(HostileKeyMapping(), "cameras", "render_all")  ->  RuntimeError: no keys for you
```

## The decision the issue asked for

#1903 laid out three candidates and noted the file had no precedent to follow, because this is the one member of the family where the read was never load-bearing. Everywhere else — #1873, #1874, #1875, #1878, #1888, #1889, #1897 — the verdict depended on the read, so a read that failed *became* the verdict. Here the verdict is settled before the read happens: a mapping is refused whatever its keys say. Only the advice was unmeasurable.

**Candidate 2 — keep the verdict, degrade the remedy** — is taken, being the only one that states both what was measured and what was not:

```
render_all: cameras must be a list of names, not a mapping, got <HostileKeyMapping object ...>.
  A mapping is iterable over its keys, so its values would be discarded - pass the names as
  a list; its own keys could not be read to quote them here (RuntimeError: no keys for you).
```

Candidate 1 (answer with the read failure) drops the mapping verdict, which is the mistake the caller actually made. Candidate 3 (drop the remedy clause) omits advice on one path and says nothing about why. Both are excluded by tests, not by prose.

The cost the issue named — *a new message form* — is paid once rather than per branch. `_describe_failed_read` names the `type: text` stem `_read_name_list` was already spelling inline, so a refusal degraded here reads like every other read failure in the module, and the two cannot describe one failure two ways. `_read_to_quote` is the shared guarded read both branches now use.

## The issue described one read. There are five, across two branches.

The issue's §Scope asked the pass to check whether any other refusal in `utils.py` reads a caller value to build its *remedy*. The string branch one line up was the weaker of the two, and none of its four escapes had been filed anywhere. Measured on `e3e5e10`:

| probe | before |
|---|---|
| `HostileKeyMapping()` — mapping `__iter__` raises | `RuntimeError: no keys for you` (the filed defect) |
| `HostileCharacterStr("wrist")` — `str` subclass `__iter__` raises | **`RuntimeError: no characters for you`** |
| `HostileLengthStr("wrist")` — `str` subclass `__len__` raises | **`RuntimeError: no length for you`** |
| `UnprintableCharacterStr("w")` — characters read fine, cannot print | **`RuntimeError: no repr for you`** |
| `UndecodableBytes(b"wrist")` — `bytes` subclass overrides `decode` | **`RuntimeError: no decoding for you`** |

Four distinct shapes on one branch: `[c for c in shown]` produced the quoted characters, a second `len(shown)` counted them, a bare f-string interpolated them (so `repr` recursed into an element that cannot print — #1875 re-raised *inside* the clause quoting it), and the overridable `bytes.decode` ran before any of that.

They now come from **one** guarded read, so the quotation and the count beside it cannot describe different values — #1897's property applied to a message rather than to a verdict.

They are fixed together because they are one shape, on one function, and splitting them would have left a measured escape one branch away needing its own issue and its own round.

## Every existing message is unchanged

Asserted as **exact text** rather than as substrings, because rewriting two message branches is precisely the change that moves a word nobody looks at. Switching the character quotation to `_refusal_container_repr` is text-preserving by construction: `f"{['w', 'r']}"` and `_refusal_container_repr(['w', 'r'])` are the same string, and the count taken from the read equals `len(shown)` for any honest `str`.

## The scan the issue said was blind, and what it found

> `TestNoGuardRendersACallerValueDirectly` scans for unrendered values, not for reads, so an unguarded `list(...)` feeding a renderer is exactly what that scan is blind to — which is how this one survived #1873 and #1875.

That gap is closed structurally rather than by fixing five call sites. `TestNoGuardReadsACallerValueOutsideATry` is added beside its sibling: it scans `utils.py` for a public guard that runs the caller's own code outside a `try`, follows the value through assignments (so `shown` — the caller's `str` under another name — is covered), and stops at the helpers that read inside a `try` and hand back an ordinary value, so the fix is not reported as loudly as the defect. Six controls, including a planted defect appended to `utils.py`'s own source so an empty table cannot mean a scanner that looks at nothing.

Against the unfixed file it reports the whole defect, including the four reads the issue did not have:

```
{'name_list_error': (('shown', 'comprehension'), ('shown', 'len()'),
                     ('value', '.decode()'), ('value', 'list()'))}
```

**It also found three more guards, and those are filed as #1906 rather than fixed here.** Each `coerce_*` guard validates its value with a validator that reads it and then reads it again, unguarded, to build the floats:

| call | result |
|---|---|
| `coerce_rgba("add_object", "color", FailsOnItsSecondRead(0.1, 0.2, 0.3))` | `RuntimeError: no second read for you` |
| `coerce_rgba(...)` with **2** components — a refusal | `RuntimeError` — on the path that exists to return text |
| `coerce_size_vector(...)`, `coerce_pose_vector(...)` | `RuntimeError` |
| `name_list_error(FailsOnItsSecondRead("top", "wrist"), ...)` | `None` — the control, read once since #1897 |

They are not absorbed because two of the three reads are on the **accepted** path and build the return value rather than a message, so "a refusal must not raise" does not reach them; and fixing it properly means the shared numeric validators return the list they read, which is a signature change on guards with many callers. They are recorded in `KNOWN_MESSAGE_READS` pointing at #1906 and pinned by `TestTheCoercionsSecondReadIsAStatedBoundary`, so closing #1906 must replace that pin and empty the table in one change — the same shape #1897 used to hand #1903 over.

## The boundary pin this replaces

`test_a_mapping_whose_keys_cannot_be_read_is_a_stated_boundary` asserted the escape with `pytest.raises`. Per the premise-test guidance it is **replaced, not deleted**: it becomes `test_a_mapping_whose_keys_cannot_be_read_keeps_its_verdict`, which asserts the verdict survives and records why the boundary existed.

## Validation

Environment: hosted runner, Python 3.13.14, no GPU/EGL, `MUJOCO_GL=disable`. Optional deps partly absent, so **every number below is a delta against the unmodified base**, per the AGENTS.md guidance.

| gate | result |
|---|---|
| `ruff check strands_robots tests tests_integ` | clean (ruff 0.15.22, the pinned range) |
| `ruff format --check strands_robots tests tests_integ` | 1039 files already formatted |
| `mypy strands_robots tests tests_integ` | **13 errors on base, 13 on branch — delta zero**; the only diff is a line number shifting on a pre-existing `utils.py` error |
| `pytest tests` (full) | **893 failures on base, 893 on branch, identical set by `diff`** — all environmental (no EGL, missing optional deps) |
| affected suite (6 files touching `name_list_error`) | 429 passed on base -> **449 passed** on branch; the same 2 no-OpenGL failures on both |
| new tests pre-fix | **8 of 10 fail**, each with the exact escaping exception; the 2 that pass are the text-preservation and probe-non-vacuity controls, which must pass both ways |
| scan pre-fix | reports all four `name_list_error` reads (above) |
| `scripts/assemble_changelog.py --check` | OK (138 pending) |

## Files

- `strands_robots/utils.py` — `_describe_failed_read`, `_read_to_quote`, both `name_list_error` branches; `_read_name_list` uses the named stem instead of repeating it (text-identical, already covered by its own tests).
- `tests/test_container_refusals_render_elementwise.py` — the boundary pin replaced, plus `TestARefusalDegradesWhenItCannotQuoteWhatItRefuses` (9 tests, 5 probes).
- `tests/test_refusal_messages_never_raise.py` — the read-side scan, its table, 8 tests and 6 controls, and the #1906 boundary pin.
- `changelog.d/1903-refusal-degrades-when-it-cannot-quote.md`.

---

### §13 changelog

**Round 1** — initial submission.

**Round 2** — the two CodeQL threads. The fix itself is unchanged.

- `py/call/wrong-arguments` (alert 858, error severity, and correct): the three coercions were
  exercised through a loop over `(guard, args)` pairs, so `coerce_pose_vector`'s extra
  component-count parameter reached every call through a `*args` splat that neither a reader nor
  static analysis can check against any of the three signatures. They are now called one by one,
  each with its own arguments, and the test docstring records why. Commit `ee97a48`; thread
  resolved by the fix.
- `py/unexpected-raise-in-special-method` (alert 859, note severity): dismissed as deliberate and
  test-only, which is the disposition this rule already carries on alerts 590
  (`_HostileRobot.__getattr__`) and 852 (`GetItemOnly.__getitem__`).
  `FailsOnItsSecondNumericRead.__getitem__` has to raise something outside `LookupError`: the
  `IndexError` the query names first is what CPython's `seqiter` catches and clears to terminate
  legacy-protocol iteration, so on a `Sequence` the suggested rewrite would end the read quietly
  and leave the three `pytest.raises` pins measuring the ABC's iteration protocol instead of the
  guards' read count. The argument is in the thread and the dismissal points at it;
  `.github/codeql/codeql-config.yml` is untouched and stays at its pinned two rule ids.

Nothing else moved this round — no change to `strands_robots/utils.py`, and the validation table
above still describes the head. Required check `call-test-lint / Test and Lint` is `SUCCESS` on
`ee97a48`, both review threads are resolved, and the `CodeQL` context sits at `NEUTRAL`, which is
advisory here.

### CODE REVIEW

**Review by yinsong1986 — DISMISSED** [review]
## Summary

Closes #1903 by making `name_list_error` keep its verdict and degrade only the quotation clause when reading the caller's value raises, instead of letting the caller's `__iter__`/`__len__`/`decode`/element-`repr` escape the guard whose purpose is to answer bad input with text. The diff matches the description: two new private helpers (`_describe_failed_read`, `_read_to_quote`, neither exported), both `name_list_error` branches rerouted through the single guarded read, `_read_name_list` switched to the shared stem (text-identical), the #1903 boundary pin replaced rather than deleted, a new AST read-side scan (`TestNoGuardReadsACallerValueOutsideATry`) with non-vacuity controls, and the three `coerce_*` second-read escapes it surfaced pinned as the #1906 boundary rather than absorbed.

## What's good

- Independently verified the text-preservation claim: `name_list_error` output is byte-identical between base `e3e5e10` and this head across honest str/bytes/mapping/scalar/sequence probes, and the five hostile probes that raised on base all return structured refusals on the branch.
- All 287 tests across the three touched test files pass on the head; `ruff check` and `ruff format --check` clean on the changed files.
- Boundary discipline follows AGENTS.md: the premise pin is replaced with a test recording why the boundary existed, and the out-of-scope `coerce_*` reads are measured, filed (#1906 exists and is open), and pinned so closing #1906 must empty `KNOWN_MESSAGE_READS` and replace the pin in one change.
- Changelog fragment present under `changelog.d/` per the fragment rule; no host paths, no emojis, no new non-ASCII in source strings.

## Verification suggestions

- Quick spot-check that the degradation is real and the honest text unmoved: `python -c "from strands_robots.utils import name_list_error as e; print(e({'top':1,'wrist':2},'cameras','render_all')); print(e('wrist','cameras','render_all'))"` and compare against the exact-text assertions in `test_no_accepted_or_refused_message_moved`.

File: tests/test_refusal_messages_never_raise.py [hunk 1]
```diff
@@ -847,6 +1011,245 @@
...
+    """
+
+    def __init__(self, *values: Any) -> None:
+        self.values = list(values)
+        self.reads = 0
+
+    def __len__(self) -> int:
+        return len(self.values)
+
+    def __getitem__(self, index: Any) -> Any:
```
[the remark this answers is not in the document]
Comment (cagataycali) [reply tests/test_refusal_messages_never_raise.py:1179]: Alert 859 is dismissed as `won't fix` (deliberate, test-only) - the same disposition this rule already carries on alerts 590 (`_HostileRobot.__getattr__`) and 852 (`GetItemOnly.__getitem__`).

`FailsOnItsSecondNumericRead` exists to answer one full read and let the next one escape, which is how the three `coerce_*` guards' unguarded second read is *measured* rather than claimed in prose (#1906, pinned by `TestTheCoercionsSecondReadIsAStatedBoundary`). The exception type is the measurement, not an oversight:

- `IndexError`, the first type this query asks for, is what CPython's `seqiter` catches and clears to terminate legacy-protocol iteration. On a `Sequence`, raising it would end the read quietly, the guard would see a short sequence instead of an escape, and all three `pytest.raises` assertions would fail - or, on a future edit, pass vacuously while measuring the ABC's iteration protocol instead of a read count. This is the failure mode AGENTS.md records from #1890, where the same suggestion on the same rule would have left a fixture asserting nothing.
- `KeyError` is the mapping form; on an integer index it models a different failure from the one #1906 turns on, and the assertions match on the message text (`no second read for you`) that the escaping read carries.

So the disposition here is a dismissal with the argument recorded, per the CI Security Baseline: rewriting a hostile fixture to satisfy a note-severity quality query is what #1879 spent a round on, and the filter file stays at its pinned two rule ids. Resolving so the merge is not held on it.

**Review by yinsong1986 — APPROVED** [review]
## Summary

Closes #1903 by making `name_list_error` keep its verdict and degrade only the quotation clause when reading the caller's value raises. Two private helpers (`_describe_failed_read`, `_read_to_quote`) centralise the failed-read spelling and the quote-only read; both `name_list_error` branches route through them, which also closes four unfiled escapes on the string branch (`__iter__`, `__len__`, element `repr`, overridable `bytes.decode`) that shared the same shape as the filed mapping-key read. Verified locally against the head: both changed test suites pass (238 tests), and `name_list_error` output is byte-identical to base for honest inputs including empty `str`/`bytes`/mapping edge cases, so the text-preservation claim holds. The change touches only private helpers and internal message construction - no public signature, wire format, or persisted schema moves.

## What's good

- The boundary pin is replaced rather than deleted, per the premise-test guidance in AGENTS.md, and the new pin records why the boundary existed.
- The three `coerce_*` second-read escapes the new scan surfaced are pinned as a measured boundary pointing at #1906 instead of being absorbed into an unrelated diff - correct scope discipline for a change whose accepted-path reads need a signature decision of their own.
- Round 2 resolved both CodeQL threads exactly as described: the splatted coercion calls are now explicit (alert 858 fixed by the diff), and the alert-859 dismissal follows the established disposition for deliberate test-only hostile fixtures (alerts 590, 852).
- Changelog fragment present and named per `changelog.d/README.md`; no host paths, no non-ASCII in the diff.
